### PR TITLE
Fix OAuth for python 3.9

### PR DESCRIPTION
**Pull request by howufiger7**

### CODE REVIEW

**Review by earonesty — APPROVED** [review]

**Review by earonesty — APPROVED** [review]

File: cloudsync/oauth/redir_server.py [hunk 1]
```diff
@@ -156,6 +156,14 @@ def shutdown(self):
             self.__on_failure = None
         self.__thread = None
 
+    def server_close(self):
+        """Closes the server and joins all threads"""
+        if self.__api_server and not self.__running:
```
Comment (earthman1) [inline cloudsync/oauth/redir_server.py:161]: If `not self.__running`, is the best behavior is to silently ignore a call to `server_close` ?

File: cloudsync/provider.py [hunk 2]
```diff
@@ -245,6 +245,8 @@ def authenticate(self) -> Creds:
                 log.error("oauth error %s", repr(e))
                 self.disconnect()
                 raise CloudTokenError(repr(e))
+            finally:
+                self._oauth_config.server_close()
```
Comment (earthman1) [inline cloudsync/provider.py:249]: Furthering the earlier comment about `server_close`, is the server stopped at this point?

File: cloudsync/tests/test_oauth_redir_server.py [hunk 3]
```diff
@@ -54,9 +71,31 @@ def send_req():
 
     shutdown_signal.wait()
     srv.shutdown()
+    srv.server_close()
+    assert server_close_calls == 1
```
Comment (earthman1) [inline cloudsync/tests/test_oauth_redir_server.py:75]: To confirm, the `EventApiServer.server_close` method gets called synchronously, and not in another thread, so this won't be timing sensitive?

File: cloudsync/provider.py (hunk 2 above)
Comment (howufiger7) [reply cloudsync/provider.py:249]: it is shutdown if wait_auth() returns / there are no exceptions.

if there is an exception, we are no worse off than before.

**Review by earthman1 — APPROVED** [review]

File: cloudsync/tests/test_oauth_redir_server.py (hunk 3 above)
Comment (howufiger7) [reply cloudsync/tests/test_oauth_redir_server.py:75]: correct. actually they are both synchronous in this test. `shutdown()` can be called from a server thread - that is covered in test_oauth.py

File: cloudsync/oauth/redir_server.py (hunk 1 above)
Comment (howufiger7) [reply cloudsync/oauth/redir_server.py:161]: moved discussion to VFM-10044

File: cloudsync/provider.py (hunk 2 above)
Comment (howufiger7) [reply cloudsync/provider.py:249]: moved discussion to VFM-10044

**Pull request merged**